cql3/expr/term_expr.cc

```diff
@@ -0,0 +1,1004 @@
+ */
+
+#include "expression.hh"
+#include "cql3/functions/function_call.hh"
+#include "cql3/column_identifier.hh"
+#include "cql3/constants.hh"
+#include "cql3/abstract_marker.hh"
+#include "cql3/lists.hh"
+#include "cql3/sets.hh"
+#include "cql3/user_types.hh"
+#include "cql3/tuples.hh"
+#include "types/list.hh"
+#include "types/set.hh"
+#include "types/map.hh"
+#include "types/user.hh"
+
+#include <boost/range/algorithm/count.hpp>
+
+namespace cql3::expr {
+
+static
+lw_shared_ptr<column_specification>
+usertype_field_spec_of(const column_specification& column, size_t field) {
+    auto&& ut = static_pointer_cast<const user_type_impl>(column.type);
+    auto&& name = ut->field_name(field);
+    auto&& sname = sstring(reinterpret_cast<const char*>(name.data()), name.size());
+    return make_lw_shared<column_specification>(
+                                   column.ks_name,
+                                   column.cf_name,
+                                   ::make_shared<column_identifier>(column.name->to_string() + "." + sname, true),
+                                   ut->field_type(field));
+}
+
+static
+void
+usertype_constructor_validate_assignable_to(const usertype_constructor& u, database& db, const sstring& keyspace, const column_specification& receiver) {
+    if (!receiver.type->is_user_type()) {
+        throw exceptions::invalid_request_exception(format("Invalid user type literal for {} of type {}", receiver.name, receiver.type->as_cql3_type()));
+    }
+
+    auto ut = static_pointer_cast<const user_type_impl>(receiver.type);
+    for (size_t i = 0; i < ut->size(); i++) {
+        column_identifier field(to_bytes(ut->field_name(i)), utf8_type);
+        if (!u.elements.contains(field)) {
+            continue;
+        }
+        const expression& value = *u.elements.at(field);
+        auto&& field_spec = usertype_field_spec_of(receiver, i);
+        if (!assignment_testable::is_assignable(test_assignment(value, db, keyspace, *field_spec))) {
+            throw exceptions::invalid_request_exception(format("Invalid user type literal for {}: field {} is not of type {}", receiver.name, field, field_spec->type->as_cql3_type()));
+        }
+    }
+}
+
+static
+assignment_testable::test_result
+usertype_constructor_test_assignment(const usertype_constructor& u, database& db, const sstring& keyspace, const column_specification& receiver) {
+    try {
+        usertype_constructor_validate_assignable_to(u, db, keyspace, receiver);
+        return assignment_testable::test_result::WEAKLY_ASSIGNABLE;
+    } catch (exceptions::invalid_request_exception& e) {
+        return assignment_testable::test_result::NOT_ASSIGNABLE;
+    }
+}
+
+static
+shared_ptr<term>
+usertype_constructor_prepare_term(const usertype_constructor& u, database& db, const sstring& keyspace, const column_specification_or_tuple& receiver_) {
+    auto& receiver = std::get<lw_shared_ptr<column_specification>>(receiver_);
```

Comment:
See a509c0f634fafacd48407628f7d489cd5be6089e

Reply:
I did, but I still don't understand.  a509c0f unifies prepare() methods from two old base classes (ostensibly "because we'll only have one expression type to represent both", which I don't think is even true) whereas this is a free function, unburdened by either inheritance or overloads.

Reply:
> I did, but I still don't understand. [a509c0f](https://github.com/scylladb/scylla/commit/a509c0f634fafacd48407628f7d489cd5be6089e) unifies prepare() methods from two old base classes (ostensibly "because we'll only have one expression type to represent both", which I don't think is even true)

Why is it not true? We have expressions that represent scalars and tuples, in a single type.

column_specification is not exactly an expression - it's a named variable, which can be either real or manufactured (the formal argument of a function). We have the real variant of this as expressions column_value and column_value_tuple. We don't yet have a way to represent a manufactured variable, but bind_variable is pretty close.

(I'd like to eliminate column_value_tuple in favor of tuple_constuctor{{column_value, column_value...}} but that's separate)

> whereas this is a free function, unburdened by either inheritance or overloads.

a509c0f weakened the type of the receiver/receivers parameter, but kept the function a member function. This commit just changes a member function to a free function.

We can't split it to two functions with separate signatures, because the caller has the unified signature.

Ah, I think I see what you mean. multi_column_restriction.hh now creates a tuple-style receiver and directly calls prepare_term (probably the only one that does this). So we could have a parallel prepare_term implementation that accepts a tuple-style receiver, instead of folding column_specification/tuple into one type and unfolding it later. This would result in some amounts of code duplication, but probably tolerable.

Ultimately, I think the way to remove column_specification_or_tuple is to make column_specification contain an expression, and pass just a column_specification (not _or_tuple). So the _or_tupleness is determined by the receiver having CQL type tuple, rather than having an overload that takes tuple, or (as is the case now) a variant that contains either a scalar or a tuple.

Reply:
To put it into concrete terms: multi_column_raw is checking that a bunch of terms are compatible with a bunch of columns:


    (x, y, z) = (2, 'rsf', 54)

the column_specification_or_tuple is the (x, y, z). expr::expression can already express (x, y, z), but it can't express typed formal variables yet. Perhaps the way to eliminate column_specification is to introduce a formal_variable expression (a type + name).

Reply:
I looked at the code, and the amount of duplication will be minimal. So the question is whether we believe in the unification for column_specification[_or_tuple] into expressions or not.

Reply:
> Ah, I think I see what you mean. multi_column_restriction.hh now creates a tuple-style receiver and directly calls prepare_term (probably the only one that does this). So we could have a parallel prepare_term implementation that accepts a tuple-style receiver, instead of folding column_specification/tuple into one type and unfolding it later.

Yes, that's what I meant.

> So the question is whether we believe in the unification for column_specification[_or_tuple] into expressions or not.

It seems clear that in this PR, the answer is still "no."

Reply:
Obviously the series doesn't do that. The question is about the direction. If it's a good direction, then splitting prepare_term() into prepare_term() and prepare_term_multi_column() will be wasted work since we'll have to reunify them (replacing the column_specification_or_tuple with a single expression). If it's not a good direction, then splitting prepare_term() removes nasty variants and improves the code.

Reply:
Seems to me like column_specification is little more than a type.  It's just used to test assignability.  I can't even find any column-name usage, outside error messages.

If types aren't expressions, neither are column_specifications.

Reply:
It's like a function formal parameter. A type accompanied by a name which is only good for error reporting.

A formal parameter isn't an expression, but it becomes part of a fake expression when the compiler assigns it an actual argument.

Right now column_specification_or_tuple can denote
 - a function formal parameter
 - a table column
 - a tuple of table columns
 - a parameter to a CQL clause (LIMIT ...)
 - tuple elements
 - user defined type element

But, it's probably faster for me to split prepare_term() than to continue arguing, and reunify it when the time comes.

Reply:
Actually, I think splitting prepare_term() can reduce duplication! Let me see if it works out.

Reply:
Done, the two new patches in the end.